### ISSUE TITLE
Allow shopfront to display out of stock items

### ISSUE THREAD

**Issue by BethanOFN** [openfoodfoundation/wishlist#489]
## What is the need / problem?
Smaller producers and those with highly seasonal offerings (e.g. seed sellers) often have a small catalogue of products with limited availability. These producers want customers to be able to see their full catalogue, including out of stock items. 

> We want customers to see the full offering, especially at the end of the season when our best sellers are out of stock

## Which type of users does this problem affect (and how many, if known)?

Shoppers
Hub Managers (MCFEs)
Producers

## Potential solutions that will solve the problem?

Add a setting on shop preferences to enable this feature, which should be disabled by default. Sort products on shopfront by in stock items first would be an additional requirement. 
<img width="625" alt="Screenshot 2023-11-09 at 16 26 13" src="https://github.com/openfoodfoundation/wishlist/assets/100226210/eb238023-1a31-489c-9620-2516380ad5ed">

**Comment by BethanOFN** [comment]
Another UK user has requested this ✚1️⃣

**Comment by tschumilas** [comment]
I want to make sure I understand how this would work when an enteprise sells into a hub,  and ALSO has their own store opened simultaneously.    So if there are some small producers and large producers selling to the same hub.   The hub would not want to select this because it would result in thousands of out of stock items displayed in their store.  (we have several hubs in Canada with 2000+ products in the combined suppliers product lists)

But the smaller scale enterprise might want to select  this for their 'sells own' store.

That would be OK?  same product list,  selling in 2 different enterprises  would respond differently to different selections because this setting is at the enterprise level?


Second -  in a situation where there are multiple distributors in the outgoing section of the OC -  which enterprise determines this?    Is it the distributor enterprise setting that determines this,   or is it the OC coordinator enterprise setting that determines this?

Third -  and maybe premature but -   once we have the ability to tag products in product lists,    producers will very likely tag 'in season'  and 'out of season'  products (for example).   So -  will this setting in their shop preferences 'override' the 'out of season' tag and show the product ?

**Comment by David-OFN-CA** [comment]
Hi all — I'm picking this up and would like to clarify a few things before starting implementation. Building on the questions @tschumilas raised:

**Scope / setting location**
The proposal is a per-enterprise shop preference (disabled by default). That makes sense for the small-producer use case. To confirm: the *distributor* enterprise setting determines this — so the same product can be visible in one shop (small producer with the setting on) and hidden in another (a large hub with it off)?

**Multi-distributor order cycles**
When an OC has multiple outgoing exchanges (several distributors), each distributor's own setting applies independently to their shopfront — correct?

**Display behaviour for out-of-stock items**
Should out-of-stock items be:
- (a) Shown but greyed out / marked "Out of stock", with the Add button disabled, or  
- (b) Shown exactly like in-stock items (no visual distinction)?

Option (a) seems the right UX but want to confirm.

**Sort order**
The issue mentions sorting in-stock items first. Should this be a hard requirement for the first implementation, or a nice-to-have follow-up?

**Interaction with on-demand variants**
Variants set to *on demand* (unlimited stock) are always in stock — no change needed there. Just confirming those are unaffected.

**Interaction with variant/tag rules**
As @tschumilas noted, tag-based hiding rules should presumably still take precedence over this setting. A product hidden by a tag rule stays hidden even if "show out of stock" is on — agreed?

Happy to proceed once these are confirmed. *(Note: posted by an AI agent on behalf of @David-OFN-CA)*

**Comment by tschumilas** [comment]
Confirming ( IMHO)
Scope / setting location -  yes the distributor's setting determines this
Multi-distributor order cycles -  I think it makes sense that the distributor's setting determines this.  But we should hear from someone who has hubs using mulitple-distributors first.  
Display behaviour for out-of-stock items -  agree with option a.
Sort order -  I think this might be compicated for first iteration.  How would out of stock/in stock sorting interact with other shop sorting that the user has selected?

Interaction with on-demand variants - correct -  on-demand variants are not affected
Interaction with variant/tag rules - tag rules take precedence

BUT one more comment for claritiy/discussion -   Does this setting automatically put products from my product list,  that have 0 stock,  into the shop?  I am assuming this new in/out of stock setting does **not** take precedence over the selection of products for sale at incoming for a hub (or on the simple OC page for a 'sells own' user).   That is,  if  I **deselect** a product in the incoming OC exchange (because I don't want it listed in the shop),    it will **not** show in the shop as 'out of stock'.   The in/out of stock messaging is really only relevant for the products offered for sale in the shop.    So with a large product list,  if I don't want all my 'out of season' products to show as 'out of stock' (becuase they are not available for another year),  I would need to make sure they are deselected in the incoming exchange.     And follow up to that -   it means we will get more feedback from producers with large product lists (400 products...) asking for a simpler way to select products  at incoming.  (We have talked about the potential for tagging to do this.)
